packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+---
+flowchart TD
+  node0((Person))
+  subgraph field0[Sequence]
+  node1((GroceryList))
+  end
+  node0--todo -->field0
+  subgraph field1[Sequence]
+  node2((Dog))
+  node3((Spider))
+  node4((Cat))
+  end
+  node0--"pets"-->field1
+  subgraph field2[Sequence]
+  node5(("String
+  ('Cat food')"))
+  node6(("String
+  ('Dog food')"))
+  end
+  node1--items-->field2
+  subgraph field3[Optional]
+  node7(("String
+  ('Bowser')"))
+  end
+  node2--name-->field3
+  subgraph field4["Optional
+  (empty)"]
+  end
+  node4--name-->field4
+```
+
+Now we can see what the data is describing - a person, including their todo list and all their pets.
+Apparently, they still need to name their cat.
+
+---
+
+This tree now has a schema and is merely one of the many possible trees that conforms to that schema.
+What is the schema in the abstract?
+Fundamentally, it's a collection of schema for each type of node.
+
+```mermaid
+---
+title: Example Schema
+---
+flowchart TD
+  node0((Person))
+  field0["Sequence of GroceryLists and TaskLists"]
+  node0--todo -->field0
+  node1((GroceryList))
+  field1["Sequence of Dogs and Cats and Birds and Spiders"]
+  node7((TaskList))
+  node7--items-->field6
+  node2((Dog))
+  node4((Cat))
+  node6((Bird))
+  node3((Spider))
+  node5(("String
+  (string data)"))
+  node0--"pets"-->field1
+  field2[Sequence of Strings]
+  node1--items-->field2
+  field3[Optional String]
+  node2--name-->field3
+  field4[Optional String]
+  node4--name-->field4
+  field5[Optional String]
+  node6--name-->field5
+  field6[Sequence of Strings]
+```
+
+This captures each of the four aspects discussed above.
+Each node (defined by its type) is associated with its expected fields and their respective kinds (multiplicities) and allowed child types.
+The example tree above is a "Person" node.
+It is just one of many possible trees that could satisfy one of the nodes in this schema.
+In fact, this schema even contains some node types that were not present in the tree at all("TaskList" and "Bird").
+
+## The Evolution of Schema
+
+Sometimes a schema needs to _change_.
+In fact, it's common for an application schema to change over time as the application matures.
+For example, new features often require new data to be stored in the tree, requiring a change to the schema.
+First, we'll identify what the possible changes to a hypothetical schema might be.
+Then, we'll explore how these changes can be accomplished by the developer.
+
+The following list describes ways that a developer might wish to change their schema.
+
+1. **Add an allowed type.**
+   A developer can expand the set of node types allowed under a particular field, or at the root of the tree.
+
+   > In fact, by modeling the tree as having a single, special root _field_ rather than having a root _node_, there is no distinction between adding a new allowed type at the root field vs. any other field in the tree.
+1. **Remove an allowed type.**
+1. **Change the type of a node.**
```

Comment:
The example of a field going from containing dogs to cats is confusing; was this just a side effect of a rename? Or did you chang the type of the field (if so, why wasn't that just a case of 1 + 2?)?